### PR TITLE
Improve / clean up meta block code & integrity

### PR DESCRIPTION
**Pull request by pdillinger**
Summary:
* Checksums are now checked on meta blocks unless specifically
suppressed or not applicable (e.g. plain table). (Was other way around.)
This means a number of cases that were not checking checksums now are,
including direct read TableProperties in Version::GetTableProperties
(fixed in meta_blocks ReadTableProperties), reading any block from
PersistentCache (fixed in BlockFetcher), read TableProperties in
SstFileDumper (ldb/sst_dump/BackupEngine) before table reader open,
maybe more.
* For that to work, I moved the global_seqno+TableProperties checksum
logic to the shared table/ code, because that is used by many utilies
such as SstFileDumper.
* Also for that to work, we have to know when we're dealing with a block
that has a checksum (trailer), so added that capability to Footer based
on magic number, and from there BlockFetcher.
* Knowledge of trailer presence has also fixed a problem where other
table formats were reading blocks including bytes for a non-existant
trailer--and awkwardly kind-of not using them, e.g. no shared code
checking checksums. (BlockFetcher compression type was populated
incorrectly.) Now we only read what is needed.
* Minimized code duplication and differing/incompatible/awkward
abstractions in meta_blocks.{cc,h} (e.g. SeekTo in metaindex block
without parsing block handle)
* Moved some meta block handling code from table_properties*.*
* Moved some code specific to block-based table from shared table/ code
to BlockBasedTable class. The checksum stuff means we can't completely
separate it, but things that don't need to be in shared table/ code
should not be.
* Use unique_ptr rather than raw ptr in more places. (Note: you can
std::move from unique_ptr to shared_ptr.)

Without enhancements to GetPropertiesOfAllTablesTest (see below),
net reduction of roughly 100 lines of code.

Test Plan: existing tests and
* Enhanced DBTablePropertiesTest.GetPropertiesOfAllTablesTest to verify that
checksums are now checked on direct read of table properties by TableCache
(new test would fail before this change)
* Also enhanced DBTablePropertiesTest.GetPropertiesOfAllTablesTest to test
putting table properties under old meta name
* Also generally enhanced that same test to actually test what it was
supposed to be testing already, by kicking things out of table cache when
we don't want them there.

### CODE REVIEW

File: table/meta_blocks.cc [hunk 1]
```diff
@@ -393,101 +404,91 @@ Status ReadProperties(const ReadOptions& read_options,
...
+  meta_index_iter->Seek(meta_block_name);
+  if (meta_index_iter->status().ok()) {
+    if (meta_index_iter->Valid() && meta_index_iter->key() == meta_block_name) {
+      Slice v = meta_index_iter->value();
+      return block_handle->DecodeFrom(&v);
+    } else if (meta_block_name == kPropertiesBlock) {
+      // Have to try old name for compatibility
+      meta_index_iter->Seek(kPropertiesBlockOldName);
+      if (meta_index_iter->status().ok() && meta_index_iter->Valid() &&
+          meta_index_iter->key() == meta_block_name) {
```
Comment (ajkr) [inline table/meta_blocks.cc:442]: Do you mean to check `key() == kPropertiesBlockOldName`?

**Review by ajkr — APPROVED** [review]
LGTM!

**Review by mrambacher — APPROVED** [review]
Seems reasonable.

File: table/block_based/block_based_table_reader.cc [hunk 2]
```diff
@@ -1852,9 +1804,9 @@ void BlockBasedTable::RetrieveMultipleBlocks(
         // begin address of each read request, we need to add the offset
         // in each read request. Checksum is stored in the block trailer,
         // beyond the payload size.
-        s = ROCKSDB_NAMESPACE::VerifyBlockChecksum(
-            footer.checksum(), data + req_offset, handle.size(),
-            rep_->file->file_name(), handle.offset());
+        s = VerifyBlockChecksum(footer.checksum(), data + req_offset,
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply table/block_based/block_based_table_reader.cc:1807]: That's in an upcoming PR actually :)

File: table/meta_blocks.cc (hunk 1 above)
Comment (pdillinger) [reply table/meta_blocks.cc:442]: Nice catch! This case probably doesn't have a unit test though I wonder if format compatibility checker goes back far enough. I should probably just use a sync point to test it.

File: table/meta_blocks.cc [hunk 3]
```diff
@@ -370,21 +372,30 @@ Status ReadProperties(const ReadOptions& read_options,
...
-      size_t len = static_cast<size_t>(handle.size() + kBlockTrailerSize);
-      *verification_buf =
-          ROCKSDB_NAMESPACE::AllocateBlock(len, memory_allocator);
-      if (verification_buf->get() != nullptr) {
-        memcpy(verification_buf->get(), block_contents.data.data(), len);
+
+  // Modified version of BlockFetcher checksum verification
+  // (See write_global_seqno comment above)
+  if (s.ok() && footer.GetBlockTrailerSize() > 0) {
+    s = VerifyBlockChecksum(footer.checksum(), block_contents.data.data(),
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply table/meta_blocks.cc:379]: Nice catch. I should not be relying on Slice moved-from semantics

**Pull request closed without merging**